src/web-vitals-autoinstrumentation.ts

```diff
@@ -0,0 +1,242 @@
+      ...this.getSharedAttributes(lcp),
+      [`${attrPrefix}.element`]: element,
+      [`${attrPrefix}.url`]: url,
+      [`${attrPrefix}.time_to_first_byte`]: timeToFirstByte,
+      [`${attrPrefix}.resource_load_delay`]: resourceLoadDelay,
+      [`${attrPrefix}.resource_load_time`]: resourceLoadTime,
+      [`${attrPrefix}.element_render_delay`]: elementRenderDelay,
+    });
+
+    span.end();
+  };
+
+  onReportINP = (inp: INPMetricWithAttribution) => {
+    const { name, attribution } = inp;
+    const { eventTarget, eventType, loadState }: INPAttribution = attribution;
+    const attrPrefix = this.getAttrPrefix(name);
+
+    const span = this.tracer.startSpan(name);
+
+    span.setAttributes({
+      ...this.getSharedAttributes(inp),
+      [`${attrPrefix}.element`]: eventTarget,
+      [`${attrPrefix}.event_type`]: eventType,
+      [`${attrPrefix}.load_state`]: loadState,
+    });
+
+    span.end();
+  };
+
+  onReportFCP = (fcp: FCPMetricWithAttribution) => {
+    const { name, attribution } = fcp;
+    const { timeToFirstByte, firstByteToFCP, loadState }: FCPAttribution =
+      attribution;
+    const attrPrefix = this.getAttrPrefix(name);
+
+    const span = this.tracer.startSpan(name);
+
+    span.setAttributes({
+      ...this.getSharedAttributes(fcp),
+      [`${attrPrefix}.time_to_first_byte`]: timeToFirstByte,
+      [`${attrPrefix}.time_since_first_byte`]: firstByteToFCP,
+      [`${attrPrefix}.load_state`]: loadState,
+    });
+
+    span.end();
+  };
+
+  onReportFID = (fid: FIDMetricWithAttribution) => {
+    const { name, attribution } = fid;
+    const { eventTarget, eventType, loadState }: FIDAttribution = attribution;
+    const attrPrefix = this.getAttrPrefix(name);
+
+    const span = this.tracer.startSpan(name);
+
+    span.setAttributes({
+      ...this.getSharedAttributes(fid),
+      [`${attrPrefix}.element`]: eventTarget,
+      [`${attrPrefix}.event_type`]: eventType,
+      [`${attrPrefix}.load_state`]: loadState,
+    });
+
+    span.end();
+  };
+
+  onReportTTFB = (ttfb: TTFBMetricWithAttribution) => {
+    const { name, attribution } = ttfb;
+    const {
+      waitingTime,
+      dnsTime,
+      connectionTime,
+      requestTime,
+    }: TTFBAttribution = attribution;
+    const attrPrefix = this.getAttrPrefix(name);
+
+    const span = this.tracer.startSpan(name);
+
+    span.setAttributes({
+      ...this.getSharedAttributes(ttfb),
+      [`${attrPrefix}.waiting_time`]: waitingTime,
+      [`${attrPrefix}.dns_time`]: dnsTime,
+      [`${attrPrefix}.connection_time`]: connectionTime,
+      [`${attrPrefix}.request_time`]: requestTime,
+    });
+
+    span.end();
+  };
+
+  enable(): void {
+    this._diag.debug(`Sending spans for ${this.vitalsToTrack.join(',')}`);
+
+    if (this.vitalsToTrack.includes('CLS')) {
+      onCLS((vital) => {
+        this.onReportCLS(vital);
+      });
+    }
+
+    if (this.vitalsToTrack.includes('LCP')) {
+      onLCP((vital) => {
+        this.onReportLCP(vital);
```

Comment:
I went back and forth a lot on whether to make these separate `onReport` functions, or whether to make a single generic one. I found that the types and tests worked out a lot better with them all being more specific.